// @nitedani/angular-renderer-core/lib/shared/angular/wrapper.js
import { Component, ViewChild, ViewContainerRef } from '@angular/core';
import * as i0 from "@angular/core";
const _c0 = ["page"];
function DefaultWrapper_ng_template_0_Template(rf, ctx) { }
export class DefaultWrapper {
    page;
    static ɵfac = function DefaultWrapper_Factory(t) { return new (t || DefaultWrapper)(); };
    static ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: DefaultWrapper, selectors: [["ng-component"]], viewQuery: function DefaultWrapper_Query(rf, ctx) { if (rf & 1) {
            i0.ɵɵviewQuery(_c0, 7, ViewContainerRef);
        } if (rf & 2) {
            let _t;
            i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.page = _t.first);
        } }, standalone: true, features: [i0.ɵɵStandaloneFeature], decls: 2, vars: 0, consts: [["page", ""]], template: function DefaultWrapper_Template(rf, ctx) { if (rf & 1) {
            i0.ɵɵtemplate(0, DefaultWrapper_ng_template_0_Template, 0, 0, "ng-template", null, 0, i0.ɵɵtemplateRefExtractor);
        } }, encapsulation: 2 });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(DefaultWrapper, [{
        type: Component,
        args: [{
                standalone: true,
                template: `<ng-template #page></ng-template>`,
            }]
    }], null, { page: [{
            type: ViewChild,
            args: ['page', { static: true, read: ViewContainerRef }]
        }] }); })();
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassDebugInfo(DefaultWrapper, { className: "DefaultWrapper", filePath: "shared/angular/wrapper.ts", lineNumber: 11 }); })();
